lucene/analysis/common/src/java/org/apache/lucene/analysis/pt/RSLPStemmerBase.java

```diff
@@ -0,0 +1,344 @@
+      for (int i = 0; i < exceptions.length; i++) {
+        if (!exceptions[i].endsWith(suffix))
+          throw new RuntimeException("warning: useless exception '" + exceptions[i] + "' does not end with '" + suffix + "'");
+      }
+      this.exceptions = new char[exceptions.length][];
+      for (int i = 0; i < exceptions.length; i++)
+        this.exceptions[i] = exceptions[i].toCharArray();
+    }
+    
+    @Override
+    public boolean matches(char s[], int len) {
+      if (!super.matches(s, len))
+        return false;
+      
+      for (int i = 0; i < exceptions.length; i++)
+        if (endsWith(s, len, exceptions[i]))
+          return false;
+
+      return true;
+    }
+  }
+  
+  /**
+   * A step containing a list of rules.
+   */
+  protected static class Step {
+    protected final String name;
+    protected final Rule rules[];
+    protected final int min;
+    protected final char[][] suffixes;
+    
+    /**
+     * Create a new step
+     * @param name Step's name.
+     * @param rules an ordered list of rules.
+     * @param min minimum word size. if this is 0 it is automatically calculated.
+     * @param suffixes optional list of conditional suffixes. may be null.
+     */
+    public Step(String name, Rule rules[], int min, String suffixes[]) {
+      this.name = name;
+      this.rules = rules;
+      if (min == 0) {
+        min = Integer.MAX_VALUE;
+        for (Rule r : rules)
+          min = Math.min(min, r.min + r.suffix.length);
+      }
+      this.min = min;
+      
+      if (suffixes == null || suffixes.length == 0) {
+        this.suffixes = null;
+      } else {
+        this.suffixes = new char[suffixes.length][];
+        for (int i = 0; i < suffixes.length; i++)
+          this.suffixes[i] = suffixes[i].toCharArray();
+      }
+    }
+    
+    /**
+     * @return new valid length of the string after applying the entire step.
+     */
+    public int apply(char s[], int len) {
+      if (len < min)
+        return len;
+      
+      if (suffixes != null) {
+        boolean found = false;
+        
+        for (int i = 0; i < suffixes.length; i++)
+          if (endsWith(s, len, suffixes[i])) {
+            found = true;
+            break;
+          }
+        
+        if (!found) return len;
+      }
+      
+      for (int i = 0; i < rules.length; i++) {
+        if (rules[i].matches(s, len))
+          return rules[i].replace(s, len);
+      }
+      
+      return len;
+    }
+  }
+  
+  /**
+   * Parse a resource file into an RSLP stemmer description.
+   * @return a Map containing the named Steps in this description.
+   */
+  protected static Map<String,Step> parse(Class<? extends RSLPStemmerBase> clazz, String resource) {
+    // TODO: this parser is ugly, but works. use a jflex grammar instead.
+    try {
+      InputStream is = clazz.getResourceAsStream(resource);
+      LineNumberReader r = new LineNumberReader(new InputStreamReader(is, StandardCharsets.UTF_8));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *r*. The resource is referenced in statements at the following lines: 249, 252, 253. The resource closure statement is at line: 256. There are other execution paths that do not contain closure statements. Either a) close *r* in a try-finally block or b) close the resource by declaring *r* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).